File: Solution MaxHeap.java

import java.util.Collections;
import java.util.PriorityQueue;

public class Solution {

    private static final int ALPHABET_SIZE = 26;

    public int minDeletions(String inputString) {
        int[] frequency = new int[ALPHABET_SIZE];
        for (int i = 0; i < inputString.length(); ++i) {
            ++frequency[inputString.charAt(i) - 'a'];
        }

        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for (int i = 0; i < ALPHABET_SIZE; ++i) {
            if (frequency[i] > 0) {
                maxHeap.add(frequency[i]);
            }
        }

        int countDeletions = 0;
        while (maxHeap.size() > 1) {
            int maxFrequency = maxHeap.poll();

            if (maxFrequency == maxHeap.peek()) {
                ++countDeletions;
                if (maxFrequency - 1 > 0) {
                    maxHeap.add(maxFrequency - 1);
                }
            }
        }
        return countDeletions;
    }
}

File: Solution MaxHeap.cpp

#include <string>
#include <vector>
using namespace std;

class Solution {
    
    inline static const int ALPHABET_SIZE = 26;

public:
    int minDeletions(const string& inputString) {

        array<int, ALPHABET_SIZE> frequency{};
        for (int i = 0; i < inputString.length(); ++i) {
            ++frequency[inputString[i] - 'a'];
        }

        priority_queue<int> maxHeap;
        for (int i = 0; i < ALPHABET_SIZE; ++i) {
            if (frequency[i] > 0) {
                maxHeap.push(frequency[i]);
            }
        }

        int countDeletions = 0;
        while (maxHeap.size() > 1) {
            int maxFrequency = maxHeap.top();
            maxHeap.pop();

            if (maxFrequency == maxHeap.top()) {
                ++countDeletions;
                if (maxFrequency - 1 > 0) {
                    maxHeap.push(maxFrequency - 1);
                }
            }
        }
        return countDeletions;
    }
};
